File: 网站后端/ChineseScoringModel/src/main/java/fun/redamancyxun/chinese/backend/score/generate_quantitative_xlsx.py
import os
import xml.etree.ElementTree as ET
import pandas as pd
import openpyxl

audio_text_mapping = {}
# workbook = openpyxl.load_workbook('src/main/java/fun/redamancyxun/chinese/backend/score/AudioTextMapping.xlsx')
workbook = openpyxl.load_workbook('/score/AudioTextMapping.xlsx')
sheet = workbook.active

# 跳过表头
for row in sheet.iter_rows(min_row=2, values_only=True):
    pcm_path, text = row
    audio_text_mapping[pcm_path] = text

for filename in audio_text_mapping.keys():

    filename = (os.path.splitext(filename)[0] + '.xml').replace('pcm', 'xml')

    print("正在处理" + filename)

    # 创建空的DataFrame来存储当前文件的数据
    all_data = []

    try:
        # 解析 XML 文件
        tree = ET.parse(filename)
        root = tree.getroot()

        # 遍历 XML 树
        for read_sentence in root.findall('.//read_sentence'):
            if not read_sentence.get('content'):
                continue
            read_sentence_data = {
                'filename': filename,
                'content': read_sentence.get('content'),
                # 'type': 'total',#
                'beg_pos': read_sentence.get('beg_pos'),
                'end_pos': read_sentence.get('end_pos'),
                # 'time_len': read_sentence.get('time_len'),
                'fluency_score': read_sentence.get('fluency_score'),
                'integrity_score': read_sentence.get('integrity_score'),
                'phone_score': read_sentence.get('phone_score'),
                'tone_score': read_sentence.get('tone_score'),
                'total_score': read_sentence.get('total_score')
            }
            all_data.append(read_sentence_data)

            for sentence in read_sentence.findall('.//sentence'):
                # sentence_data = {
                #     'filename': filename,
                #     'content': sentence.get('content'),
                #     'type': 'sentence',
                #     'beg_pos': sentence.get('beg_pos'),
                #     'end_pos': sentence.get('end_pos'),
                #     # 'time_len': sentence.get('time_len'),
                #     'phone_score': '',
                #     'tone_score': '',
                #     'total_score': ''
                # }
                # all_data.append(sentence_data)

                for word in sentence.findall('.//word'):
                    sylls = word.findall('.//syll')
                    for syll in sylls:
                        # syll_data = {
                        #     'filename': filename,
                        #     'content': syll.get('content'),
                        #     'type': 'syll',
                        #     'beg_pos': syll.get('beg_pos'),
                        #     'end_pos': syll.get('end_pos'),
                        #     'time_len': syll.get('time_len'),
                        #     'dp_message': syll.get('dp_message'),
                        #     'rec_node_type': syll.get('rec_node_type'),
                        #     'symbol': syll.get('symbol'),
                        # }
                        # all_data.append(syll_data)

                        phones = syll.findall('.//phone')
                        for phone in phones:
                            phone_data = {
                                'filename': filename,
                                'content': phone.get('content'),
                                # 'type': 'phone',
                                'beg_pos': phone.get('beg_pos'),
                                'end_pos': phone.get('end_pos'),
                                # 'time_len': phone.get('time_len'),
                                'dp_message': phone.get('dp_message'),
                                # 'rec_node_type': phone.get('rec_node_type'),#
                                # 'symbol': '',
                                'mono_tone': phone.get('mono_tone'),
                                'is_yun': phone.get('is_yun'),
                                'perr_msg': phone.get('perr_msg'),
                                # 'perr_level_msg': phone.get('perr_level_msg')
                            }
                            all_data.append(phone_data)

        # 将当前文件的数据存储到DataFrame中
        df = pd.DataFrame(all_data)

        # 将DataFrame保存为Excel文件，文件名与XML文件一致
        df.to_excel(filename.replace('.xml', ".xlsx").replace('xml', 'xlsx'), index=False)

    except Exception as e:
        print(f"处理文件 {filename} 时出错: {e}")

File: 网站后端/ChineseScoringModel/src/main/java/fun/redamancyxun/chinese/backend/score/GenerateQuantitativeXlsx.java
package fun.redamancyxun.chinese.backend.score;

import org.apache.commons.io.FilenameUtils;
import org.apache.poi.ss.usermodel.*;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GenerateQuantitativeXlsx {

    public static void main(String[] args) {
        try {
            // Load Excel file
            FileInputStream excelFile = new FileInputStream(new File("src/main/java/fun/redamancyxun/chinese/backend/score/AudioTextMapping.xlsx"));
            Workbook workbook = new XSSFWorkbook(excelFile);
            Sheet sheet = workbook.getSheetAt(0);

            // Map PCM paths to text
            Map<String, String> audioTextMapping = new HashMap<>();
            for (Row row : sheet) {
                if (row.getRowNum() == 0) continue; // Skip header
                String pcmPath = row.getCell(0).getStringCellValue();
                String text = row.getCell(1).getStringCellValue();
                audioTextMapping.put(pcmPath, text);
            }

            // Process each XML file
            for (String pcmPath : audioTextMapping.keySet()) {
                String xmlFilename = (FilenameUtils.removeExtension(pcmPath) + ".xml").replace("pcm", "xml");
                System.out.println("正在处理 " + xmlFilename);

                // Parse XML file
                File xmlFile = new File(xmlFilename);
                DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
                DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
                Document doc = dBuilder.parse(xmlFile);

                // Normalize the document
                doc.getDocumentElement().normalize();

                // Collect data from XML
                List<Map<String, String>> allData = new ArrayList<>();
                NodeList readSentenceNodes = doc.getElementsByTagName("read_sentence");
                for (int temp = 0; temp < readSentenceNodes.getLength(); temp++) {
                    Element readSentence = (Element) readSentenceNodes.item(temp);
                    if (readSentence.getAttribute("content").isEmpty()) continue;

                    Map<String, String> readSentenceData = new HashMap<>();
                    readSentenceData.put("filename", xmlFilename);
                    readSentenceData.put("content", readSentence.getAttribute("content"));
                    readSentenceData.put("beg_pos", readSentence.getAttribute("beg_pos"));
                    readSentenceData.put("end_pos", readSentence.getAttribute("end_pos"));
                    readSentenceData.put("fluency_score", readSentence.getAttribute("fluency_score"));
                    readSentenceData.put("integrity_score", readSentence.getAttribute("integrity_score"));
                    readSentenceData.put("phone_score", readSentence.getAttribute("phone_score"));
                    readSentenceData.put("tone_score", readSentence.getAttribute("tone_score"));
                    readSentenceData.put("total_score", readSentence.getAttribute("total_score"));
                    allData.add(readSentenceData);

                    NodeList sentenceNodes = readSentence.getElementsByTagName("sentence");
                    for (int i = 0; i < sentenceNodes.getLength(); i++) {
                        Element sentence = (Element) sentenceNodes.item(i);
                        NodeList wordNodes = sentence.getElementsByTagName("word");
                        for (int j = 0; j < wordNodes.getLength(); j++) {
                            Element word = (Element) wordNodes.item(j);
                            NodeList syllNodes = word.getElementsByTagName("syll");
                            for (int k = 0; k < syllNodes.getLength(); k++) {
                                Element syll = (Element) syllNodes.item(k);
                                NodeList phoneNodes = syll.getElementsByTagName("phone");
                                for (int l = 0; l < phoneNodes.getLength(); l++) {
                                    Element phone = (Element) phoneNodes.item(l);
                                    Map<String, String> phoneData = new HashMap<>();
                                    phoneData.put("filename", xmlFilename);
                                    phoneData.put("content", phone.getAttribute("content"));
                                    phoneData.put("beg_pos", phone.getAttribute("beg_pos"));
                                    phoneData.put("end_pos", phone.getAttribute("end_pos"));
                                    phoneData.put("dp_message", phone.getAttribute("dp_message"));
                                    phoneData.put("mono_tone", phone.getAttribute("mono_tone"));
                                    phoneData.put("is_yun", phone.getAttribute("is_yun"));
                                    phoneData.put("perr_msg", phone.getAttribute("perr_msg"));
                                    allData.add(phoneData);
                                }
                            }
                        }
                    }
                }

                // Save data to Excel
                Workbook excelWorkbook = new XSSFWorkbook();
                Sheet excelSheet = excelWorkbook.createSheet("Data");
                int rowCount = 0;
                for (Map<String, String> data : allData) {
                    Row row = excelSheet.createRow(rowCount++);
                    int colCount = 0;
                    for (String key : data.keySet()) {
                        Cell cell = row.createCell(colCount++);
                        cell.setCellValue(data.get(key));
                    }
                }

                try (FileOutputStream outputStream = new FileOutputStream((FilenameUtils.removeExtension(xmlFilename) + ".xlsx").replace("xml", "xlsx"))) {
                    excelWorkbook.write(outputStream);
                }

            }

        } catch (IOException | ParserConfigurationException | SAXException e) {
            e.printStackTrace();
        }
    }

}
